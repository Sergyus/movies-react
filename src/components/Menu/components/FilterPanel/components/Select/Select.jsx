import React, { useCallback, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import styles from '../../FilterPanel.scss';

const Select = ({
  setActive, setActiveItem, fetchedGenres, fetchMoviesByGenre, filter,
}) => {
  const [selectValue, setSelectValue] = useState('');
  const genres = fetchedGenres;
  const genreList = genres.map(item => (
    <option value={item.id} key={item.id}>{item.name}</option>
  ));
  useEffect(() => {
    if (filter !== 'Genres') {
      setSelectValue('');
    }
  }, [filter, setSelectValue]);
  return (
    <li className={styles.navItem}>
      <select
        onChange={useCallback((e) => {
          setSelectValue(e.target.value);
          setActive('Genres');
          fetchMoviesByGenre(1, e.target.value);
        }, [setActive, fetchMoviesByGenre, setSelectValue])}
        aria-label="Genres"
        value={selectValue}
        className={cn(styles.selectGenre, setActiveItem('Genres'))}
        name="selectGenre"
      >
        <option value="" disabled hidden>Genre</option>
        { genreList }
      </select>
    </li>
  );
};

Select.propTypes = {
  setActive: PropTypes.func,
  setActiveItem: PropTypes.func,
  fetchedGenres: PropTypes.arrayOf(PropTypes.object),
  fetchMoviesByGenre: PropTypes.func,
  filter: PropTypes.string,
};

Select.defaultProps = {
  setActive: () => {},
  setActiveItem: () => {},
  fetchedGenres: [],
  fetchMoviesByGenre: () => {},
  filter: 'Trending',
};

export default Select;
