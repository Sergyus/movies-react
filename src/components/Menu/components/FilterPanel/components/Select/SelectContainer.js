import { connect } from 'react-redux';
import Select from './Select';
import { fetchMoviesByGenre, fetchedGenresSelector, filterSelector } from '../../../../../../modules/app';

const mapStateToProps = state => ({
  fetchedGenres: fetchedGenresSelector(state),
  filter: filterSelector(state),
});

export default connect(mapStateToProps, { fetchMoviesByGenre })(Select);
