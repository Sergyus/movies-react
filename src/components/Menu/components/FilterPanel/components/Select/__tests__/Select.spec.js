import React from 'react';
import { create, act } from 'react-test-renderer';
import Select from '../Select';
import { genres } from '../../../../../../../assets/json/genres.json';

describe('Select component', () => {
  const setActiveItem = jest.fn();
  const setActive = jest.fn();
  const fetchMoviesByGenre = jest.fn();
  it('Select component renders correctly', () => {
    const tree = create(
      <Select
        activeItem="Trending"
        setActiveItem={setActiveItem}
        fetchedGenres={genres}
        setActive={setActive}
        fetchMoviesByGenre={fetchMoviesByGenre}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
  it('Select component with filter renders correctly', () => {
    let tree;
    act(() => {
      tree = create(
        <Select filter="Genres" />,
      );
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });
  it('Select component without props works correctly', () => {
    const tree = create(
      <Select />,
    );
    const select = tree.root.findByProps({ 'aria-label': 'Genres' });
    const mockOnChange = jest.fn(select.props.onChange);
    act(() => {
      mockOnChange({ target: { value: '123' } });
    });
    expect(mockOnChange).toHaveBeenCalled();
  });
});
