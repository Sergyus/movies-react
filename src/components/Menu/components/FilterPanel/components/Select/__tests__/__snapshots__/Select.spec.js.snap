// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component Select component renders correctly 1`] = `
<li
  className="navItem"
>
  <select
    aria-label="Genres"
    className="selectGenre"
    name="selectGenre"
    onChange={[Function]}
    value=""
  >
    <option
      disabled={true}
      hidden={true}
      value=""
    >
      Genre
    </option>
    <option
      value={28}
    >
      Action
    </option>
    <option
      value={12}
    >
      Adventure
    </option>
    <option
      value={16}
    >
      Animation
    </option>
    <option
      value={35}
    >
      Comedy
    </option>
    <option
      value={80}
    >
      Crime
    </option>
    <option
      value={99}
    >
      Documentary
    </option>
    <option
      value={18}
    >
      Drama
    </option>
    <option
      value={10751}
    >
      Family
    </option>
    <option
      value={14}
    >
      Fantasy
    </option>
    <option
      value={36}
    >
      History
    </option>
    <option
      value={27}
    >
      Horror
    </option>
    <option
      value={10402}
    >
      Music
    </option>
    <option
      value={9648}
    >
      Mystery
    </option>
    <option
      value={10749}
    >
      Romance
    </option>
    <option
      value={878}
    >
      Science Fiction
    </option>
    <option
      value={10770}
    >
      TV Movie
    </option>
    <option
      value={53}
    >
      Thriller
    </option>
    <option
      value={10752}
    >
      War
    </option>
    <option
      value={37}
    >
      Western
    </option>
  </select>
</li>
`;

exports[`Select component Select component with filter renders correctly 1`] = `
<li
  className="navItem"
>
  <select
    aria-label="Genres"
    className="selectGenre"
    name="selectGenre"
    onChange={[Function]}
    value=""
  >
    <option
      disabled={true}
      hidden={true}
      value=""
    >
      Genre
    </option>
  </select>
</li>
`;
